x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/action/apikey/RestGrantApiKeyAction.java

```diff
@@ -42,38 +46,65 @@
  */
 @ServerlessScope(Scope.INTERNAL)
 public final class RestGrantApiKeyAction extends ApiKeyBaseRestHandler implements RestRequestFilter {
+    public interface RequestTranslator {
+        GrantApiKeyRequest translate(RestRequest request) throws IOException;
 
-    static final ObjectParser<GrantApiKeyRequest, Void> PARSER = new ObjectParser<>("grant_api_key_request", GrantApiKeyRequest::new);
-    static {
-        PARSER.declareString((req, str) -> req.getGrant().setType(str), new ParseField("grant_type"));
-        PARSER.declareString((req, str) -> req.getGrant().setUsername(str), new ParseField("username"));
-        PARSER.declareField(
-            (req, secStr) -> req.getGrant().setPassword(secStr),
-            SecurityBaseRestHandler::getSecureString,
-            new ParseField("password"),
-            ObjectParser.ValueType.STRING
-        );
-        PARSER.declareField(
-            (req, secStr) -> req.getGrant().setAccessToken(secStr),
-            SecurityBaseRestHandler::getSecureString,
-            new ParseField("access_token"),
-            ObjectParser.ValueType.STRING
-        );
-        PARSER.declareString((req, str) -> req.getGrant().setRunAsUsername(str), new ParseField("run_as"));
-        PARSER.declareObject(
-            (req, clientAuthentication) -> req.getGrant().setClientAuthentication(clientAuthentication),
-            CLIENT_AUTHENTICATION_PARSER,
-            new ParseField("client_authentication")
-        );
-        PARSER.declareObject(
-            (req, api) -> req.setApiKeyRequest(api),
-            (parser, ignore) -> CreateApiKeyRequestBuilder.parse(parser),
-            new ParseField("api_key")
-        );
+        class Default implements RequestTranslator {
```

Comment:
The reason I didn't pull this into its own class is cumbersome dependencies: `SecurityBaseRestHandler::getSecureString` is protected, and so is `CLIENT_AUTHENTICATION_PARSER`. I could move things around but I don't think that's currently worth it.

I think the numerous extension points (and the number of different exported modules) is not ideal, and I'd like to take some time to think about how to consolidate them. Perhaps the right answer is to actually move all translators from various other places to the Rest action classes, like we have it here. I need to stare at it a little longer though to decide.